platform/logger.h

```diff
@@ -0,0 +1,105 @@
+
+#include <stdio.h>
+#include <stdarg.h>
+#include "platform/mbed_preprocessor.h"
+
+#ifndef MBED_CONF_MAX_LOG_STR_SIZE
+#define MBED_CONF_MAX_LOG_STR_SIZE          128
+#endif
+
+#ifndef MBED_CONF_LOG_MAX_BUFFER_SIZE
+#define MBED_CONF_LOG_MAX_BUFFER_SIZE       1024
+#endif
+
+#define LOG_TOTAL_BUFFER_SIZE       MBED_CONF_LOG_MAX_BUFFER_SIZE
+#define LOG_SINGLE_STR_SIZE         MBED_CONF_MAX_LOG_STR_SIZE
+
+#if defined(__ARMCC_VERSION)
+#define FILE_NAME_                  __MODULE__
+#define FILE_INFO_                  __MODULE__ " " MBED_STRINGIFY(__LINE__) " "
+#else
+#define FILE_NAME_                  __BASE_FILE__
+#define FILE_INFO_                  __BASE_FILE__ " " MBED_STRINGIFY(__LINE__) " "
+#endif
+
+#define LOG_LEVEL_ERR_CRITICAL      0x1
+#define LOG_LEVEL_ERR               0x2
+#define LOG_LEVEL_WARN              0x4
+#define LOG_LEVEL_INFO              0x8
+#define LOG_LEVEL_DEBUG             0x10
+
+// Log-Level Strings
+#define LOG_ERR_CRITICAL_        "CRT"
+#define LOG_ERR_                 "ERR"
+#define LOG_WARN_                "WRN"
+#define LOG_DEBUG_               "DBG"
+#define LOG_INFO_                "INF"
+
+#define GET_MODULE_ID_(a, b, c, d)   (375 + (a << 1) + (b << 2) + (c >> 2) + (d >> 1))
+#define GET_MODULE_ID(x)        (GET_MODULE_ID_(x[0], x[1], x[2], x[3]))
+
+#define SET_MODULE(x)           (GET_MODULE_ID(x) & 0xFF)
+#define SET_COUNTER(y)          ((y & 0xFF) << 8)
+#define SET_LINE_NUM(z)         ((z & 0xFFFF) << 16)
+#define TRACE_ID_(x,y,z)        (SET_MODULE(x) | SET_COUNTER(y) | SET_LINE_NUM(z))
+
+typedef struct trace_id {
+    uint32_t t_id;
+    uint32_t strLen;
+    char *s;
+}trace_id_t;
+
+// Macros to log ID based data
+#define MBED_LOG_ID_4(...)                           log_buffer_id_data(__VA_ARGS__)
+#define MBED_LOG_ID_3(counter, id, args, fmt, ...)   ({volatile static const __attribute__((section(".keep.log_data"))) char str##counter[] = fmt; \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t len##counter = MBED_STRLEN(fmt); \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t c##counter = id; \
+                                                       MBED_LOG_ID_4(args, id, ##__VA_ARGS__); \
+                                                      })
+#define MBED_LOG_ID_2(counter, id, ...)               MBED_LOG_ID_3(counter, id, MBED_COUNT_VA_ARGS(__VA_ARGS__), ##__VA_ARGS__)
+#define MBED_LOG_ID_1(mod, fmt, ll, f, l, c, ...)     MBED_LOG_ID_2(c, TRACE_ID_(mod,c,l), mod " " f " " MBED_STRINGIFY(l) " : " fmt "\n", ##__VA_ARGS__)
+                                                      
+// Macros to log string data
+#define MBED_LOG_STR(...)                          log_buffer_string_data(__VA_ARGS__)
+#define MBED_LOG_STR_1(mod, fmt, ll, f, l, ...)    MBED_LOG_STR("%-3.3s %-4.4s %-15s %5d : " fmt "\n", ll, mod, f, l, ##__VA_ARGS__)
```

Comment:
Do you really need to be outputting line numbers and filenames? And consuming all that ROM to store them?

This is general tracing, not asserts. The "module" and content of the string should be easily enough to locate it. Asserts need the file + line number because they don't usually contain any text, just the condition.

If you do want to output filename+line (make it configurable?), at least scrap the `FILE_INFO` concept. By concatenating the filename and number as a string, that means you're requiring the compiler to store a separate copy of the filename for every single line.

Given how much ROM this string tracing must be using, I can kind of understand why you'd want to go for ID-based logging.

Reply:
Also, you can save some space by leaving the `\n` off. You are assuming one line per call, so no need to embed it in the callsite.

As to the line prefix, can't you format them internally in your library, as mbed_trace does?

You're building into a line buffer anyway, so can do it as two vsnprintfs - one for the fixed parameters and one for the user format. That also permits verbosity/style changes easily.

Basically, you're doing the same thing as mbed trace, but a bit less efficiently?

Reply:
> As to the line prefix, can't you format them internally in your library, as mbed_trace does?

Compile time macros are used to switch between string based and ID based tracing, where we want to replace all the strings. Instead of having multiple strings ID, we would like to have single string with all necessary info easily replaced by single ID.

Reply:
> You're building into a line buffer anyway, so can do it as two vsnprintfs - one for the fixed parameters and one for the user format.

We are not creating line buffer, we are using circular buffer which cannot differentiate between multiple lines

Reply:
You get your debug function to output the line separator (or but a separator into your buffer) itself.  You don't need to put the `\n` in the caller's string. You know it's a line because that's part of the API - one call = one line.